File: bubble_sort.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <chrono>
#include <iomanip>
#include <sstream>
struct SortStep {
    int step;
    std::vector<int> array;
    std::vector<int> comparing;
    bool swapped;
    std::string description;
    
    SortStep(int s, const std::vector<int>& arr, const std::vector<int>& comp, 
             bool swap, const std::string& desc) 
        : step(s), array(arr), comparing(comp), swapped(swap), description(desc) {}
};

std::vector<SortStep> bubbleSortWithSteps(std::vector<int>& arr) {
    std::vector<SortStep> steps;
    int n = arr.size();
    
    // Record initial state
    steps.emplace_back(0, arr, std::vector<int>{}, false, "Initial array");
    
    int stepCount = 1;
    
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n - i - 1; j++) {
            // Record comparison step
            steps.emplace_back(stepCount, arr, std::vector<int>{j, j + 1}, false,
                "Comparing elements at indices " + std::to_string(j) + " and " + std::to_string(j + 1));
            stepCount++;
            
            // If elements need to be swapped
            if (arr[j] > arr[j + 1]) {
                std::swap(arr[j], arr[j + 1]);
                
                // Record swap step
                steps.emplace_back(stepCount, arr, std::vector<int>{j, j + 1}, true,
                    "Swapped elements at indices " + std::to_string(j) + " and " + std::to_string(j + 1));
                stepCount++;
            }
        }
    }
    
    // Record final state
    steps.emplace_back(stepCount, arr, std::vector<int>{}, false, "Array sorted!");
    
    return steps;
}

void saveToJson(const std::vector<SortStep>& steps, double executionTime, const std::vector<int>& originalArray) {
    std::ofstream file("cpp_steps.json");
    
    if (!file.is_open()) {
        std::cerr << "Error: Could not open file for writing!" << std::endl;
        return;
    }
    
    file << "{\n";
    file << "  \"language\": \"C++\",\n";
    file << "  \"algorithm\": \"Bubble Sort\",\n";
    file << "  \"original_array\": [";
    for (size_t i = 0; i < originalArray.size(); i++) {
        file << originalArray[i];
        if (i < originalArray.size() - 1) file << ", ";
    }
    file << "],\n";
    file << "  \"execution_time\": " << std::fixed << std::setprecision(6) << executionTime << ",\n";
    file << "  \"total_steps\": " << steps.size() << ",\n";
    file << "  \"steps\": [\n";
    
    for (size_t i = 0; i < steps.size(); i++) {
        const auto& step = steps[i];
        file << "    {\n";
        file << "      \"step\": " << step.step << ",\n";
        file << "      \"array\": [";
        for (size_t j = 0; j < step.array.size(); j++) {
            file << step.array[j];
            if (j < step.array.size() - 1) file << ", ";
        }
        file << "],\n";
        file << "      \"comparing\": [";
        for (size_t j = 0; j < step.comparing.size(); j++) {
            file << step.comparing[j];
            if (j < step.comparing.size() - 1) file << ", ";
        }
        file << "],\n";
        file << "      \"swapped\": " << (step.swapped ? "true" : "false") << ",\n";
        file << "      \"description\": \"" << step.description << "\"\n";
        file << "    }";
        if (i < steps.size() - 1) file << ",";
        file << "\n";
    }
    
    file << "  ]\n";
    file << "}\n";
    
    file.close();
}

int main() {
    // Shared input array - same for all languages
    std::vector<int> inputArray = {64, 34, 25, 12, 22, 11, 90, 88, 76, 50, 42};
    
    std::cout << "C++ - Original array: ";
    for (size_t i = 0; i < inputArray.size(); i++) {
        std::cout << inputArray[i];
        if (i < inputArray.size() - 1) std::cout << " ";
    }
    std::cout << std::endl;
    
    auto startTime = std::chrono::high_resolution_clock::now();
    std::vector<SortStep> steps = bubbleSortWithSteps(inputArray);
    auto endTime = std::chrono::high_resolution_clock::now();
    
    auto duration = std::chrono::duration_cast<std::chrono::nanoseconds>(endTime - startTime);
    double executionTime = duration.count() / 1e9;
    
    std::cout << "C++ - Sorted array: ";
    for (size_t i = 0; i < steps.back().array.size(); i++) {
        std::cout << steps.back().array[i];
        if (i < steps.back().array.size() - 1) std::cout << " ";
    }
    std::cout << std::endl;
    
    std::cout << "C++ - Execution time: " << std::fixed << std::setprecision(6) << executionTime << " seconds" << std::endl;
    std::cout << "C++ - Total steps: " << steps.size() << std::endl;
    
    // Save steps to JSON file
    saveToJson(steps, executionTime, {64, 34, 25, 12, 22, 11, 90, 88, 76, 50, 42});
    
    std::cout << "C++ - Steps saved to cpp_steps.json" << std::endl;
    
    return 0;
}

File: bubble_sort.py
import json
import time

def bubble_sort_with_steps(arr):
    """
    Performs bubble sort and records each step of the sorting process
    """
    steps = []
    n = len(arr)
    start_time = time.time()
    
    # Record initial state
    steps.append({
        "step": 0,
        "array": arr.copy(),
        "comparing": [],
        "swapped": False,
        "description": "Initial array"
    })
    
    step_count = 1
    
    for i in range(n):
        for j in range(0, n - i - 1):
            # Record comparison step
            steps.append({
                "step": step_count,
                "array": arr.copy(),
                "comparing": [j, j + 1],
                "swapped": False,
                "description": f"Comparing elements at indices {j} and {j + 1}"
            })
            step_count += 1
            
            # If elements need to be swapped
            if arr[j] > arr[j + 1]:
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                
                # Record swap step
                steps.append({
                    "step": step_count,
                    "array": arr.copy(),
                    "comparing": [j, j + 1],
                    "swapped": True,
                    "description": f"Swapped elements at indices {j} and {j + 1}"
                })
                step_count += 1
    
    end_time = time.time()
    execution_time = end_time - start_time
    
    # Record final state
    steps.append({
        "step": step_count,
        "array": arr.copy(),
        "comparing": [],
        "swapped": False,
        "description": "Array sorted!"
    })
    
    return steps, execution_time

def main():
    # Shared input array - same for all languages
    input_array = [64, 34, 25, 12, 22, 11, 90, 88, 76, 50, 42]
    print(f"Python - Original array: {input_array}")
    
    # Perform bubble sort and get steps
    steps, exec_time = bubble_sort_with_steps(input_array.copy())
    
    print(f"Python - Sorted array: {steps[-1]['array']}")
    print(f"Python - Execution time: {exec_time:.6f} seconds")
    print(f"Python - Total steps: {len(steps)}")
    
    # Save steps to JSON file
    output_data = {
        "language": "Python",
        "algorithm": "Bubble Sort",
        "original_array": [64, 34, 25, 12, 22, 11, 90, 88, 76, 50, 42],
        "execution_time": exec_time,
        "total_steps": len(steps),
        "steps": steps
    }
    
    with open("python_steps.json", "w") as f:
        json.dump(output_data, f, indent=2)
    
    print("Python - Steps saved to python_steps.json")

if __name__ == "__main__":
    main()